### PR TITLE
fix(webpack-dev-server): do not encodeUri in loader

### PR DESCRIPTION
**Pull request by lmiller1990**
- Closes https://github.com/cypress-io/cypress/issues/20573

### User facing changelog

Correctly load specs with `[` and `]` characters when using `@cypress/webpack-dev-server`.

### Additional details

Not really relevant to people outside Cypress, but when you back merge this to 10.0-release, you can use this branch for an example (same changes but updated system-tests project: https://github.com/cypress-io/cypress/compare/10.0-release...lmiller1990/20573?expand=1).

### How has the user experience changed?

N/A

### PR Tasks

- [x] Have tests been added/updated?
- [ ] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [ ] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: system-tests/test/webpack_dev_server_spec.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import systemTests from '../lib/system-tests'
+
+describe('@cypress/webpack-dev-server', function () {
+  systemTests.setup()
```
Comment (lmiller1990) [inline system-tests/test/webpack_dev_server_spec.ts:4]: The start of the desperately needed system tests for WDS.

File: npm/webpack-dev-server/src/loader.ts [hunk 2]
```diff
@@ -18,7 +18,7 @@ const makeImport = (file: Cypress.Cypress['spec'], filename: string, chunkName:
   const magicComments = chunkName ? `/* webpackChunkName: "${chunkName}" */` : ''
 
   return `"${filename}": {
-    shouldLoad: () => document.location.pathname.includes("${encodeURI(file.absolute)}"),
```
Comment (lmiller1990) [inline npm/webpack-dev-server/src/loader.ts:21]: I guess we don't need this. If it passes CI, I'm pretty confident we are good to go.

**Comment by emilyrohrbough** [comment]
@lmiller1990 Does this close https://github.com/cypress-io/cypress/issues/5426 as well?

**Comment by ZachJW34** [comment]
@emilyrohrbough This is a CT specific change, so no this shouldn't fix the issue coming from e2e tests.

**Review by ZachJW34 — COMMENTED** [review]
@lmiller1990 @JessicaSachs tested this with a filename with a space and confirmed it does not work. `[]` are valid in some URLs. We can change the `shouldLoad` to something like
```
// Uses the browser URL parsing method, so if it does change per browser we would still match
shouldLoad: () => document.location.pathname.includes("${new URL(`http://localhost${file.absolute}`).pathname}")

// alternative: see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURI#encoding_for_ipv6
function fixedEncodeURI(str) {
    return encodeURI(str).replace(/%5B/g, '[').replace(/%5D/g, ']');
}
shouldLoad: () => document.location.pathname.includes("${fixedEncodeURI(file.absolute)}"),
```

Test cases covering both brackets and encoded characters would be great.

**Comment by JessicaSachs** [comment]
Why wouldn't we just decodeURI at the other end of whatever is consuming this?

**Comment by ZachJW34** [comment]
@JessicaSachs We consume it in the same line, so we could do 
```
shouldLoad: () => decodeUri(document.location.pathname).includes(encodeUri(file.absolute))
```

Haven't tested it but should work.

**Comment by ZachJW34** [comment]
@lmiller1990 created a PR targeting your branch that use `decodeURI` that will support both escaped characters and `[]`
https://github.com/cypress-io/cypress/pull/20593

File: npm/webpack-dev-server/src/loader.ts (hunk 2 above)
Comment (lmiller1990) [reply npm/webpack-dev-server/src/loader.ts:21]: We already had spaces: https://github.com/cypress-io/cypress/pull/20575/files#diff-f5c1d2cf3b3dd9c9d3c0464a216d71fdf03609a366cfe48406b699823ddce7f7R138

- `...`: https://github.com/cypress-io/cypress/pull/20575/files#diff-f5c1d2cf3b3dd9c9d3c0464a216d71fdf03609a366cfe48406b699823ddce7f7R119
- unicode: https://github.com/cypress-io/cypress/pull/20575/files#diff-f5c1d2cf3b3dd9c9d3c0464a216d71fdf03609a366cfe48406b699823ddce7f7R100

File: packages/server/lib/controllers/iframes.ts [hunk 3]
```diff
@@ -39,7 +39,7 @@ export const iframesController = {
     // attach header data for webservers
     // to properly intercept and serve assets from the correct src root
     // TODO: define a contract for dev-server plugins to configure this behavior
-    req.headers.__cypress_spec_path = req.params[0]
+    req.headers.__cypress_spec_path = encodeURI(req.params[0])
```
Comment (lmiller1990) [inline packages/server/lib/controllers/iframes.ts:42]: Needed this guy, too.

**Comment by lmiller1990** [comment]
@JessicaSachs @ZachJW34 This is ready to go, it has tests for all the suggested edge cases (unicode, spaces, dots, square brackets). We needed to add this line on the server: https://github.com/cypress-io/cypress/pull/20575/files#diff-f41297d10b0dfd5080e119a99dbb460a6ac6defccb4cf402b189c2e1e3b96947R42.

I am not sure how this works with `10.0-release`, we will find out when we rebase. The tests should help.

**Review by ZachJW34 — APPROVED** [review]
We should have similar tests for the vite-dev-server. I think it was confirmed that it's working but would be nice to have the same coverage across both dev-servers. Not relevant for this PR though!

File: packages/server/lib/controllers/iframes.ts (hunk 3 above)
Comment (lmiller1990) [reply packages/server/lib/controllers/iframes.ts:42]: Yes, when we request a spec from the AUT it goes through this controller on the way to webpack-dev-server.

**Comment by JessicaSachs** [comment]
> We should have similar tests for the vite-dev-server. I think it was confirmed that it's working but would be nice to have the same coverage across both dev-servers. Not relevant for this PR though!

Agreed

**Review by JessicaSachs — APPROVED** [review]
Thanks for adding all of those tests @lmiller1990

**Comment by lmiller1990** [comment]
Two approvals - going to update with latest master, then merge.

**Pull request merged**